devicetypes/smartthings/zigbee-accessory-dimmer.src/zigbee-accessory-dimmer.groovy

```diff
@@ -44,62 +45,121 @@ metadata {
 
 def getSTEP() {10}
 
+def getSCENES_CLUSTER() { 0x0005 }
+def getSCENES_STORE_SCENE_COMMAND() { 0x04 }
+def getSCENES_RECALL_SCENE_COMMAND() { 0x05 }
+
+def getONOFF_ON_COMMAND() { 0x01 }
+def getONOFF_OFF_COMMAND() { 0x00 }
+def getLEVEL_MOVE_LEVEL_COMMAND() { 0x00 }
+def getLEVEL_MOVE_COMMAND() { 0x01 }
+def getLEVEL_STEP_COMMAND() { 0x02 }
+def getLEVEL_STOP_COMMAND() { 0x03 }
+def getLEVEL_MOVE_LEVEL_ONOFF_COMMAND() { 0x04 }
+def getLEVEL_MOVE_ONOFF_COMMAND() { 0x05 }
+def getLEVEL_STEP_ONOFF_COMMAND() { 0x06 }
+def getLEVEL_STOP_ONOFF_COMMAND() { 0x07 }
+def getLEVEL_DIRECTION_UP() { "00" }
+def getLEVEL_DIRECTION_DOWN() { "01" }
+
+
+def isIkeaDimmer() {
+	device.getDataValue("model") == "TRADFRI wireless dimmer"
+}
+
 // Parse incoming device messages to generate events
 def parse(String description) {
 	log.debug "description is $description"
+	def results = []
 
 	def event = zigbee.getEvent(description)
 	if (event) {
 		if (event.name=="level" && event.value==0) {}
 		else {
-			sendEvent(event)
+			results << createEvent(event)
 		}
 	} else {
 		def descMap = zigbee.parseDescriptionAsMap(description)
-		if (descMap && descMap.clusterInt == 0x0006) {
-			if (descMap.commandInt == 0x01 || descMap.commandInt == 0x00) {
+
+		if (descMap && descMap.clusterInt == zigbee.ONOFF_CLUSTER) {
```

Comment:
Could you add some comments describing the messages each of the different devices use and how they map to capabilities? Some of these changes don't seem to make sense but I think it would be a lot easier to understand with some additional context.

Reply:
@tpmanley Done! Let me know if you think a few more are needed? I didn't add a comment for the ONOFF logic since I am not sure how the Aurora remote handles it (I assume it only sends one type of packet and thus only supports a "toggle" action?), but set the default action to be to create the event based off of the command the device sends us.

Reply:
I think it's still a bit hard to understand. What do you think about breaking this into separate functions for each model? There might be a little bit of duplicated code but I think it will be a lot more readable and easier to support/extend.

Reply:
I re-organized some code that hopefully makes it a little easier. I feel like the duplicate code would be too great if I handled it per device. There are really only a couple of areas where the implementations differ.